aries_cloudagent/resolver/default/peer3.py

```diff
@@ -0,0 +1,193 @@
+                )
+        await _reset_keys_from_did_doc(profile, dp3_doc)
+
+        return dp3_doc
+
+
+async def _reset_keys_from_did_doc(profile, did_doc):
+    async with profile.session() as session:
+        storage: BaseStorage = session.inject(BaseStorage)
+        await storage.delete_all_records(
+            BaseConnectionManager.RECORD_TYPE_DID_KEY, {"did": did_doc.id}
+        )
+
+    for vm in did_doc.verification_method or []:
+        if vm.controller == did_doc.id:
+            if vm.public_key_base58:
+                await _add_key_for_did(profile, did_doc.id, vm.public_key_base58)
+            if vm.public_key_multibase:
+                pk = multibase.decode(vm.public_key_multibase)
+                if len(pk) == 32:  # No multicodec prefix
+                    pk = bytes_to_b58(pk)
+                else:
+                    codec, key = multicodec.unwrap(pk)
+                    if codec == multicodec.multicodec("ed25519-pub"):
+                        pk = bytes_to_b58(key)
+                    else:
+                        continue
+                await _add_key_for_did(profile, did_doc.id, pk)
+
+
+async def _add_key_for_did(profile, did: str, key: str):
+    """Store a verkey for lookup against a DID.
+
+    Args:
+        did: The DID to associate with this key
+        key: The verkey to be added
+    """
+    record = StorageRecord(
+        BaseConnectionManager.RECORD_TYPE_DID_KEY, key, {"did": did, "key": key}
+    )
+    async with profile.session() as session:
+        storage: BaseStorage = session.inject(BaseStorage)
+        try:
+            await storage.find_record(
+                BaseConnectionManager.RECORD_TYPE_DID_KEY, {"key": key}
+            )
+        except StorageNotFoundError:
+            await storage.add_record(record)
+        except StorageDuplicateError:
+            pass
+            # "Key already associated with DID: %s; this is likely caused by "
+            # "routing keys being erroneously stored in the past",
+
+
+def _replace_all_values(input, org, new):
+    for k, v in input.items():
+        if isinstance(v, type(dict)):
+            _replace_all_values(v, org, new)
+        if isinstance(v, List):
+            for i, item in enumerate(v):
+                if isinstance(item, type(dict)):
+                    _replace_all_values(item, org, new)
+                elif (
+                    isinstance(item, str)
+                    or isinstance(item, DID)
+                    or isinstance(item, DIDUrl)
+                ):
+                    v.pop(i)
+                    v.append(item.replace(org, new, 1))
+                elif hasattr(item, "__dict__"):
+                    _replace_all_values(item.__dict__, org, new)
+                else:
+                    pass
+
+        elif isinstance(v, str) or isinstance(v, DID) or isinstance(v, DIDUrl):
+            input[k] = v.replace(org, new, 1)
+        else:
+            pass
```

Comment:
This (and other methods) really ought to have some doc strings at least explaining their purpose and why they're necessary. This is a pretty dense method especially so having a summary in English of whats going on would save others a fair amount of time parsing through it themselves.

Reply:
Replaced with a much better way of doing this.